File: tsoc15c1p5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cstring>

using namespace std;

vector<int> G[101];
int ants[101], car[101];
bool V[101];

int main() {
    int N, M, X, Y; cin >> N >> M;
    for (int i = 0; i < M; i++) {
        cin >> X >> Y;
        G[X].push_back(Y);
        G[Y].push_back(X);
    }
    int W, disp; cin >> W;
    queue<int> Q;
    for (int i = 0; i < W; i++) {
        cin >> disp;
        Q.push(disp);
        ants[disp] = 0;
        V[disp] = true;
    }
    while (!Q.empty()) {
        int C = Q.front();
        Q.pop();
        for (int i : G[C]) {
            if (!V[i]) {
                V[i] = true;
                ants[i] = ants[C] + 4;
                Q.push(i);
            }
        }
    }
    memset(V, false, sizeof(V));
    car[1] = 0;
    V[1] = 0;
    Q.push(1);
    while (!Q.empty()) {
        int C = Q.front();
        Q.pop();
        for (int i : G[C]) {
            if (!V[i] && car[C] + 1 < ants[i]) {
                V[i] = true;
                car[i] = car[C] + 1;
                Q.push(i);
            }
        }
    }
    cout << (V[N] ? to_string(car[N]) : "sacrifice bobhob314");
    return 0;
}

File: tsoc15c1p5.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        int N = readInt(), M = readInt(), X, Y, ants[] = new int[N + 1], car[] = new int[N + 1];
        List<Integer> G[] = new ArrayList[N + 1];
        boolean V[] = new boolean[N + 1];
        for (int i = 1; i <= N; i++) G[i] = new ArrayList<>();
        for (int i = 0; i < M; i++) {
            X = readInt(); Y = readInt();
            G[X].add(Y); G[Y].add(X);
        }
        Queue<Integer> Q = new LinkedList<>();
        for (int W = readInt(); W > 0; W--) {
            int disp = readInt();
            Q.add(disp);
            V[disp] = true;
            ants[disp] = 0;
        }
        while (!Q.isEmpty()) {
            int C = Q.poll();
            for (int i : G[C]) {
                if (!V[i]) {
                    V[i] = true;
                    ants[i] = ants[C] + 4;
                    Q.add(i);
                }
            }
        }
        for (int i = 1; i <= N; i++) V[i] = false;
        V[1] = true;
        car[1] = 0;
        Q.add(1);
        while (!Q.isEmpty()) {
            int C = Q.poll();
            for (int i : G[C]) {
                if (!V[i] && ants[i] > car[C] + 1) {
                    V[i] = true;
                    car[i] = car[C] + 1;
                    Q.add(i);
                }
            }
        }
        System.out.println(V[N] ? car[N] : "sacrifice bobhob314");
    }

    static String next () throws IOException {
        while (st == null || !st.hasMoreTokens()) st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }

    static int readInt () throws IOException {
        return Integer.parseInt(next());
    }
}